File: internal/email/service.go
package email

import (
	"context"
	"fmt"
)

// InboxService coordinates polling, correlation, reply persistence, and HWM
// advancement. The HWM is staged while replies are inserted: a failed insert
// leaves the persisted HWM unchanged, so a retry can safely ingest the batch.
type InboxService struct {
	Dialer IMAPDialer
	State  HWMStore
}

func NewInboxService(dialer IMAPDialer, state HWMStore) *InboxService {
	return &InboxService{Dialer: dialer, State: state}
}

// PollAndMatch fetches and correlates replies, persists them, and commits the
// staged HWM only after every persistence operation succeeds.
func (s *InboxService) PollAndMatch(ctx context.Context, cfg IMAPConfig, folders []string, requests []RemovalRequest, threadMap map[string]int64, store ReplyStore) ([]MatchedMessage, error) {
	if s == nil || s.Dialer == nil {
		return nil, fmt.Errorf("%w: dialer is nil", ErrIMAP)
	}
	state := s.State
	if state == nil {
		state = NewMemoryHWMStore()
	}
	staged := NewStagingHWMStore(state)
	messages, err := pollFoldersWithState(ctx, cfg, folders, s.Dialer, staged)
	if err != nil {
		return nil, err
	}
	matched := MatchReplyToRequest(messages, requests, threadMap)
	if store != nil {
		for _, reply := range matched {
			if err := store.Insert(ctx, reply, ParseEmailBody(reply.Message.Body, 200)); err != nil {
				return nil, fmt.Errorf("email: persist inbox reply: %w", err)
			}
		}
	}
	if err := staged.Commit(ctx); err != nil {
		return nil, fmt.Errorf("email: commit inbox high-water mark: %w", err)
	}
	return matched, nil
}

func pollFoldersWithState(ctx context.Context, cfg IMAPConfig, folders []string, dialer IMAPDialer, state HWMStore) ([]Message, error) {
	if len(folders) == 0 {
		folders = []string{"INBOX"}
	}
	seen := make(map[string]struct{})
	var all []Message
	for _, folder := range folders {
		cfg.Folder = folder
		messages, err := pollInbox(ctx, cfg, dialer, state)
		if err != nil {
			return nil, err
		}
		for _, message := range messages {
			key := message.MessageID
			if key == "" {
				key = message.ID
			}
			if _, ok := seen[key]; ok {
				continue
			}
			seen[key] = struct{}{}
			all = append(all, message)
		}
	}
	return all, nil
}

File: crates/symeraseme-core/src/email/service.rs
//! Inbox service: poll, correlate, persist, and only then commit the
//! high-water mark. A failed insert leaves the persisted mark untouched, so a
//! retry can safely ingest the same batch again.
//!
//! Ported from `internal/email/service.go`.

use crate::email::hwm::{HwmStore, MemoryHwmStore, StagingHwmStore};
use crate::email::policy::{match_reply_to_request, parse_email_body, poll_folders};
use crate::email::session::ImapDialer;
use crate::email::types::{ImapConfig, ImapError, MatchedMessage, RemovalRequest};
use std::collections::HashMap;

/// Persistence boundary for inbox replies. Implementations store bounded
/// snippets and parsed metadata — never credentials or raw authentication
/// responses.
pub trait ReplyStore {
    fn insert(&self, reply: &MatchedMessage, snippet: &str) -> Result<(), String>;
}

pub struct InboxService<'a> {
    dialer: &'a dyn ImapDialer,
    state: Option<&'a dyn HwmStore>,
}

impl<'a> InboxService<'a> {
    pub fn new(dialer: &'a dyn ImapDialer, state: Option<&'a dyn HwmStore>) -> Self {
        Self { dialer, state }
    }

    pub fn poll_and_match(
        &self,
        config: &ImapConfig,
        folders: &[String],
        requests: &[RemovalRequest],
        thread_map: &HashMap<String, i64>,
        store: Option<&dyn ReplyStore>,
    ) -> Result<Vec<MatchedMessage>, ImapError> {
        let owned_state = MemoryHwmStore::new();
        let state: &dyn HwmStore = self.state.unwrap_or(&owned_state);
        let staged = StagingHwmStore::new(Some(state));
        let messages = poll_folders(config, folders, self.dialer, &staged)?;
        let matched = match_reply_to_request(&messages.unwrap_or_default(), requests, thread_map);
        if let Some(store) = store {
            for reply in &matched {
                store
                    .insert(reply, &parse_email_body(&reply.message.body, 200))
                    .map_err(|cause| {
                        ImapError::new(format!("email: persist inbox reply: {cause}"))
                    })?;
            }
        }
        staged.commit().map_err(|cause| {
            ImapError::new(format!("email: commit inbox high-water mark: {cause}"))
        })?;
        Ok(matched)
    }
}
